### PR TITLE
Ibm backend v2 (solves #318 and #347)

### LINKED ISSUE #318: Is there any way to access the IBM Q 14 Melbourne?
Hi,

Since the IBM Q 16 seems to be retired, is there any way to use Project Q to access and run code on the IBM Q 14 Melbourne? 

Thank you very much in advance

Best,

Elías

### LINKED ISSUE #347: Can not connect to IMBQ
When making requests in _ibm_http_client file such as in the method authenticate, it is returning an html file, not a json file which is throwing exceptions. The reason might be that the api links are not working no more.

### PR DESCRIPTION
**Pull request by dbretaud**
Happy new year everyone!

* updates on the ibm backend to access again ibmq devices (5 and 15 qubit devices + 32 qubit simulator) 
* update on ibm setup to handle the three different devices and decompose gates into available gate set (they now only accept u1,u2,u3->rotation gates ,id,barrier and cx gates)
* ibm example updated

-> not sure if we should suppress ibm16 setup or simply indicates it is deprecated. File unmodified so far

Closes #318 and closes #347

### CODE REVIEW

File: projectq/backends/_ibm/_ibm_http_client.py [hunk 1]
```diff
@@ -21,75 +21,221 @@
...
+        return device in self.backends
+
+    def can_run_experiment(self,info,device):
+        """
+        check if the device is big enough to run the code
+
+        Args:
+            info (dict): dictionary sent by the backend containing the code to run
+            device (str): name of the ibm device to use
+        :return:
```
Comment (Takishima) [inline projectq/backends/_ibm/_ibm_http_client.py]: I would rather have you follow the style used throughout ProjectQ:
```python
"""
        Returns:
```

File: projectq/backends/_ibm/_ibm_http_client.py [hunk 2]
```diff
@@ -21,75 +21,221 @@
...
+            (bool): True if device is big enough, False otherwise
```
Comment (Takishima) [inline projectq/backends/_ibm/_ibm_http_client.py]: Please also document the two other return values.

File: projectq/backends/_ibm/_ibm_http_client.py [hunk 3]
```diff
@@ -21,75 +21,221 @@
...
+        r_json=r.json()
+        self.backends=dict()
+        for el in r_json:
+            self.backends[el['backend_name']]={'nq':el['n_qubits'],'coupling_map':el['coupling_map'],'version':el['backend_version']}
+        if verbose==True:
+            print('- List of IBMQ devices available:')
+            print(self.backends)
+        return self.backends
+
+    def is_online(self,device):
```
Comment (Takishima) [inline projectq/backends/_ibm/_ibm_http_client.py]: Missing docstring

File: projectq/backends/_ibm/_ibm_http_client.py [hunk 4]
```diff
@@ -21,75 +21,221 @@
...
+CLIENT_APPLICATION = 'ibmqprovider/0.4.4'#TODO: call to get the API version automatically
+
+
+class IBMQ(Session):
+    def __init__(self):
+        super().__init__()
+        self.backends=dict()
+        self.timeout=5.0
+
+    def get_list_devices(self,verbose=False):
```
Comment (Takishima) [inline projectq/backends/_ibm/_ibm_http_client.py]: Missing docstring

File: projectq/backends/_ibm/_ibm_test.py [hunk 5]
```diff
@@ -175,7 +175,6 @@ def mock_send(*args, **kwargs):
     qureg = eng.allocate_qureg(3)
     # entangle the qureg
     Entangle | qureg
-    Tdag | qureg[0]
```
Comment (Takishima) [inline projectq/backends/_ibm/_ibm_test.py]: Any particular reason why you removed the Tdag gate from this test?

File: projectq/backends/_ibm/_ibm_http_client.py [hunk 6]
```diff
@@ -21,75 +21,221 @@
...
+def show_devices(token,verbose=False):    
+    """
+    Access the list of available devices and their properties (ex: for setup configuration)
 
-def is_online(device):
-    url = 'Backends/{}/queue/status'.format(device)
-    r = requests.get(urljoin(_api_url, url))
-    return r.json()['state']
+    Args:
+        token (str): IBM quantum experience user password.
```
Comment (Takishima) [inline projectq/backends/_ibm/_ibm_http_client.py]: I would change "user password" to "user API token" to match what is displayed on the IBMQ website.

**Comment by Takishima** [comment]
Many thanks for working on this issue and as previously: great work 😎

Did not have time to fully test the code, but from what I can see, this looks pretty good already.
I will try testing it next week and will do a final review then.

**Comment by dbretaud** [comment]
I don't have more modifications for the IBM backend, so I think it's ready for merging :)

**Pull request merged**